compiler/rustc_middle/src/middle/privacy.rs

```diff
@@ -103,31 +130,89 @@ impl<Id: Hash + Eq + Copy> AccessLevels<Id> {
         })
     }
 
+    pub fn get_effective_vis(&self, id: Id) -> Option<&EffectiveVisibility> {
+        self.map.get(&id)
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = (&Id, &EffectiveVisibility)> {
+        self.map.iter()
+    }
+
+    pub fn map_id<OutId: Hash + Eq + Copy>(&self, f: impl Fn(Id) -> OutId) -> AccessLevels<OutId> {
+        AccessLevels { map: self.map.iter().map(|(k, v)| (f(*k), *v)).collect() }
+    }
+}
+
+impl<Id: Hash + Eq + Copy + Into<DefId>> AccessLevels<Id> {
     pub fn set_access_level(&mut self, id: Id, tag: AccessLevel) {
-        let mut effective_vis = self.get_effective_vis(id).copied().unwrap_or_default();
+        // FIXME: The id in Visibility::Restricted must be a normal module ID,
+        // but this function marks items accessible to other crates and don't
+        // support EffectiveVisibility invariants
+        let mut effective_vis = self
+            .get_effective_vis(id)
+            .cloned()
+            .unwrap_or(EffectiveVisibility::from_vis(Visibility::Restricted(id.into())));
         for level in [
             AccessLevel::Public,
             AccessLevel::Exported,
             AccessLevel::Reachable,
             AccessLevel::ReachableFromImplTrait,
         ] {
             if level <= tag {
-                *effective_vis.get_mut(level) = Some(Visibility::Public);
+                *effective_vis.get_mut(level) = Visibility::Public;
             }
         }
         self.map.insert(id, effective_vis);
     }
 
-    pub fn get_effective_vis(&self, id: Id) -> Option<&EffectiveVisibility> {
-        self.map.get(&id)
-    }
-
-    pub fn iter(&self) -> impl Iterator<Item = (&Id, &EffectiveVisibility)> {
-        self.map.iter()
+    pub fn try_init(&mut self, id: Id, module_id: Option<Id>) {
```

Comment:
```suggestion
    pub fn try_init(&mut self, id: Id, module_vis: Visibility) {
```
This `Option` is equivalent to `Visibility` if you look at the body of this function (None -> Public, Some -> Restricted).

Reply:
Also it's better to add such `module_vis: Visibility` parameter to `fn set_access_level` as well.